src/parser/jobs/smn/modules/Devotion.js

```diff
@@ -0,0 +1,144 @@
+import {t} from '@lingui/macro'
+import {Trans} from '@lingui/react'
+import {getDataBy} from 'data/getDataBy'
+import Module from 'parser/core/Module'
+import DISPLAY_ORDER from './DISPLAY_ORDER'
+import ACTIONS from 'data/ACTIONS'
+import STATUSES from 'data/STATUSES'
+import {RotationTable} from 'components/ui/RotationTable'
+import React from 'react'
+
+const PLAYERS_HIT_TARGET = 8
+const MAX_DEVOTION_DURATION = 30000
+
+export default class Devotion extends Module {
+	static displayOrder = DISPLAY_ORDER.DEVOTION
+	static handle = 'devotion'
+	static title = t('smn.devotion.title')`Devotion Actions`
+	static dependencies = [
+		'combatants',
+		'timeline',
+	]
+
+	_devotionWindows = []
+
+	// {
+	//	start: Number
+	//	end: Number
+	//	playersHit: [Number]
+	//	events: [Event]
+	// }
+	_currentWindow = null
+
+	constructor(...args) {
+		super(...args)
+		this.addEventHook('cast', {by: 'player'}, this._onCast)
+		this.addEventHook('cast', {by: 'pet', abilityId: ACTIONS.DEVOTION.id}, this._onPetDevotionCast)
+		this.addEventHook('complete', this._onComplete)
+		this.addEventHook('normalisedapplybuff', {by: 'pet', abilityId: STATUSES.DEVOTION.id}, this._countDevotionTargets)
+		this.parser.fightFriendlies.forEach(id => {
+			this.addEventHook('removebuff', {to: id, abilityId: STATUSES.DEVOTION.id}, this._onDevotionRemoved)
+		})
+	}
+
+	_onComplete() {
+		this._closeWindow(this.parser.fight.end_time - this.parser.fight.start_time)
+	}
+
+	_onDevotionRemoved(event) {
+		this._closeWindow(event.timestamp)
+	}
+
+	_onPetDevotionCast(event) {
+		this._openWindow(event.timestamp)
+	}
+
+	_countDevotionTargets(event) {
+		const playersHit = event.confirmedEvents.map(hit => hit.targetID).filter(id => this.parser.fightFriendlies.findIndex(f => f.id === id) >= 0)
```

Comment:
This could prob be broken into a few lines just for readability. Can keep the method chaining, just throw each chain on a new line, like this.
```suggestion
		const playersHit = event.confirmedEvents
			.map(hit => hit.targetID)
			.filter(id => this.parser.fightFriendlies.findIndex(f => f.id === id) >= 0)
```

Reply:
These changes have all been implemented